File: 프로그래머스/2/131704. 택배상자/택배상자.py
def solution(order):
    stack = []
    answer = 0
    n = len(order)
    idx = 0  # order에서 현재 실어야 할 상자 인덱스

    for box in range(1, n+1):  # 1부터 n까지 컨베이어 벨트에서 상자 도착
        stack.append(box)  # 일단 보조 컨테이너에 넣음

        # 보조 컨테이너의 top이 현재 order와 같으면 꺼내서 트럭에 싣기
        while stack and stack[-1] == order[idx]:
            stack.pop()
            answer += 1
            idx += 1
    return answer

File: 프로그래머스/2/131704. 택배상자/택배상자.java
/*
컨테이너 벨트 : 한 방향 진행, 벨트에 놓인 순서대로 내릴 수 있음
보조 컨테이너 벨트 : 양방향 이동, 스택
*/
import java.util.*;
class Solution {
    public int solution(int[] order) {
        int answer = 0;
        int n = order.length;
        Stack<Integer> stack = new Stack<>();
        int last = 0;
        int j = 0; // 택배 순서 인덱스
        for (int i = 1; i < n+1; i++) {
            if (i == order[j]) {
                //현재 택배 인덱스와 같다면
                answer++;
                j++;
            }
            else {
                stack.push(i);
            }
            while(!stack.isEmpty() && j < n && stack.peek() == order[j]) {
                stack.pop();
                answer++;
                j++;
            }
        }
        return answer;
    }
}